File: CLIENT-JAVA/src/main/java/com/variant/client/impl/StateRequestImpl.java
package com.variant.client.impl;

import java.util.Map;
import java.util.Optional;
import java.util.Set;

import com.variant.client.Session;
import com.variant.client.StateRequest;
import com.variant.client.VariantException;
import com.variant.client.util.MethodTimingWrapper;
import com.variant.client.TraceEvent;
import com.variant.share.error.ServerError;
import com.variant.share.schema.State;
import com.variant.share.schema.StateVariant;
import com.variant.share.schema.Variation;
import com.variant.share.schema.Variation.Experience;
import com.variant.share.session.CoreStateRequest;

/**
 * Permanent wrapper around transient CoreStateRequest + client only request functionality.
 */
public class StateRequestImpl implements StateRequest {

	private final SessionImpl session;
	private final ConnectionImpl conn;
	private CoreStateRequest coreRequest;
	private StateVisitedEvent sve;
	
	/**
	 * Ok to use this object?
	 */
	private void checkState() {
		session.preChecks();
	}
	
	/**
	 * Both commit() and fail() call here.
	 * @param state
	 * @param userData
	 */
	private void _commit(StateRequest.Status status, Object... userData) {
		
		checkState();
		
		// If local state already reflects target state -- noop.
		if (status.ordinal() != coreRequest.getStatus().ordinal()) {
		
			// Persist targeting and session ID trackers.  Note that we expect the userData to apply to both.
			session.saveTrackers(userData);
			
			// Commit in shared state TODO: make this async
			conn.client.server.requestCommit(this, status);
			
			sve = null;
		}
	}

	public StateRequestImpl(SessionImpl session) 
	{	
		this.session = session;
		this.conn = (ConnectionImpl) session.getConnection();
		this.coreRequest = session.getCoreSession().getStateRequest().get();
		this.sve = new StateVisitedEvent(this.coreRequest.getState());
	}

	//---------------------------------------------------------------------------------------------//
	//                                            PUBLIC                                           //
	//---------------------------------------------------------------------------------------------//

	@Override
	public State getState() {
		checkState();
		return coreRequest.getState();
	}

	@Override
	public Optional<StateVariant> getResolvedStateVariant() {
		checkState();
		return coreRequest.getResolvedStateVariant();
	}

	@Override
	public Map<String,String> getResolvedParameters() {
		checkState();
		return coreRequest.getResolvedParameters();
	}

	@Override
	public Set<Experience> getLiveExperiences() {
		checkState();
		return coreRequest.getLiveExperiences();
	}

	@Override
	public Optional<Experience> getLiveExperience(Variation var) {
		checkState();
		return coreRequest.getLiveExperience(var);
	}

	@Override
	public void commit(Object... userData) {
		
		new MethodTimingWrapper<Object>().exec( () -> {
			if (getStatus() == StateRequest.Status.Failed)
				throw new VariantException(ServerError.CANNOT_COMMIT);
			else if (getStatus() == StateRequest.Status.InProgress)
				_commit(StateRequest.Status.Committed, userData);
			return null;  // void method -- making compiler happy.
		});
	}
	
	@Override
	public void fail(Object... userData) {

		new MethodTimingWrapper<Object>().exec( () -> {
			if (getStatus() == StateRequest.Status.Committed)
				throw new VariantException(ServerError.CANNOT_FAIL);
			else if (getStatus() == StateRequest.Status.InProgress)
				_commit(StateRequest.Status.Failed, userData);
			return null;  // void method -- making compiler happy.
		});
	}
	
	@Override
	public StateRequest.Status getStatus() {
		return StateRequest.Status.values()[coreRequest.getStatus().ordinal()];
	}

	@Override
	public TraceEvent getStateVisitedEvent() {
		return sve;
	}

	/**
	 * Override with a narrower return type to return the client session, instead of core.
	 */
	@Override
	public Session getSession() {
		return session;
	}

	// ---------------------------------------------------------------------------------------------//
	//                                           PUBLIC EXT                                         //
	// ---------------------------------------------------------------------------------------------//
	
	public void rewrap(CoreStateRequest coreRequest) {
		this.coreRequest = coreRequest;
	}

}

File: SERVER/app/com/variant/server/impl/StateRequestImpl.scala
package com.variant.server.impl

import com.variant.core.schema.State
import com.variant.server.api.StateRequest
import com.variant.core.session.CoreStateRequest
import java.util.Date
import com.variant.core.schema.Variation
import com.variant.core.schema.Variation.Experience
import com.variant.core.schema.StateVariant
import com.variant.core.TraceEvent
import com.variant.core.StateRequestStatus
import com.variant.core.schema.impl.StateVariantImpl
import com.variant.server.api.Session
import com.variant.server.event.ServerTraceEvent
import com.variant.core.impl.StateVisitedEvent

/**
 * 
 */
object StateRequestImpl {
   
   def apply(session: Session, coreReq:CoreStateRequest) = new StateRequestImpl(session, coreReq)
}

/**
 * 
 */
class StateRequestImpl(private val session: Session, private val coreReq:CoreStateRequest) extends StateRequest {
  
   /*----------------------------------------------------------------------------------------*/
   /*                                        PUBLIC                                          */
   /*----------------------------------------------------------------------------------------*/

   /**
	 */
	override def getSession: Session = session

	/**
	 */
	override def getState: State = coreReq.getState

	/**
	 */
	override def getStatus: StateRequestStatus = coreReq.getStatus

	/**
	 */
	override def getLiveExperiences: java.util.Set[Experience] = coreReq.getLiveExperiences

	/**
	 */
	override def getLiveExperience(variation: Variation): java.util.Optional[Experience] = coreReq.getLiveExperience(variation)

	/**
    */
	override def getResolvedStateVariant: java.util.Optional[StateVariant] = coreReq.getResolvedStateVariant
	
	/**
	 */	
	override def getResolvedParameters: java.util.Map[java.lang.String, java.lang.String] = coreReq.getResolvedParameters
   
   /*----------------------------------------------------------------------------------------*/
   /*                                     PUBLIC EXT                                         */
   /*----------------------------------------------------------------------------------------*/
	
   def setResolvedStateVariant(variant: java.util.Optional[StateVariant]): Unit = coreReq.setResolvedStateVariant(variant)
         
   def setStatus(status: StateRequestStatus): Unit = coreReq.setStatus(status)

}